src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+                default:
+                    return undefined;
+            }
+        }
+    }, [selectedOption, resourceType, isValidUrlStr, t]);
+
+    const isValuesEqual = (value1: string, value2: string) => {
+        if (displayField === AzureResourceDisplayFields.url) {
+            if (value1?.endsWith('/')) {
+                value1 = value1.slice(0, -1);
+            }
+            if (value2?.endsWith('/')) {
+                value2 = value2.slice(0, -1);
+            }
+            return value1 === value2;
+        } else {
+            return value1 === value2;
+        }
+    };
+
+    const handleOnChange = useCallback(
+        (option, value) => {
+            if (option) {
+                setSelectedOption(option);
+                setSelectedKey(option.data?.id ?? option.text);
+                if (onResourceChange) {
+                    onResourceChange(
+                        option.data || option.text,
+                        options
+                            ?.filter(
+                                (option) =>
+                                    option.itemType !==
+                                    SelectableOptionMenuItemType.Header
+                            )
+                            .map((option) => option.data || option.text) || []
+                    );
+                }
+            } else {
+                const newParsedOptionValue = parsedOptionText(value);
+
+                const existingOption = options.find((option) =>
+                    isValuesEqual(option.text, newParsedOptionValue)
+                );
+                if (!existingOption) {
+                    const newOption = {
+                        key: newParsedOptionValue,
+                        text: newParsedOptionValue,
+                        styles: comboBoxOptionStyles
+                    };
+
+                    let newOptions = deepCopy(options);
+                    if (
+                        displayField !== AzureResourceDisplayFields.url ||
+                        (displayField === AzureResourceDisplayFields.url &&
+                            isValidUrlStr(newParsedOptionValue))
+                    ) {
+                        const freeFromOptionsHeaderExist = options?.find(
+                            (option) =>
+                                option.itemType ===
+                                    SelectableOptionMenuItemType.Header &&
+                                option.text === freeformOptionsHeaderText
+                        );
+                        newOptions = freeFromOptionsHeaderExist
+                            ? [...newOptions, newOption]
+                            : [...newOptions, freeformOptionsHeader, newOption];
+                        setOptions(newOptions);
+                        setSelectedOption(newOption);
+                        setSelectedKey(newParsedOptionValue);
+                    }
+                    if (onResourceChange)
```

Comment:
can we pull out the duplicate code that notifies the caller of the change? It seems like both conditionals do the same thing.

Reply:
Fixed: https://github.com/microsoft/iot-cardboard-js/commit/6a3da6ace5c92b93b9f5386ad828f0e9f3326ae4#diff-3bd141423ed5699ed9c540576ed4472a3ed44ba1f9f6c4e73b720705561b84deR499